### PR TITLE
INT-4283: MarkerFilePresentFileListFilters

### PR DESCRIPTION
**Pull request by garyrussell**
JIRA: https://jira.spring.io/browse/INT-4283

### CODE REVIEW

File: spring-integration-file/src/main/java/org/springframework/integration/file/filters/AbstractMarkerFilePresentFileListFilter.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+ */
+public abstract class AbstractMarkerFilePresentFileListFilter<F> implements FileListFilter<F> {
+
+	private final Map<FileNameFileListFilter<F>, Function<String, String>> filtersAndFunctions = new HashMap<>();
+
+	/**
+	 * Construct an instance with a single {@link FileNameFileListFilter} and a suffix that will
+	 * be used to replace the file suffix of matched files when looking for the marker file.
+	 * i.e. if a file {@code foo.txt} is matched by the filter and the suffix is "complete",
+	 * this filter will only pass "foo.txt" if "foo.complete" is present.
```
Comment (artembilan) [inline spring-integration-file/src/main/java/org/springframework/integration/file/filters/AbstractMarkerFilePresentFileListFilter.java]: M-m-m. Why do you select the extension replacing approach as default one?
Seems for me `foo.txt.complete` is much robust solution, especially when we may have `foo.exe`, `foo.jpg`, `foo.java` at the same folder.

**Review by artembilan — CHANGES_REQUESTED** [review]
That's for this round

File: spring-integration-file/src/main/java/org/springframework/integration/file/filters/AbstractMarkerFilePresentFileListFilter.java (hunk 1 above)
Comment (garyrussell) [inline spring-integration-file/src/main/java/org/springframework/integration/file/filters/AbstractMarkerFilePresentFileListFilter.java]: :+1:

File: spring-integration-file/src/main/java/org/springframework/integration/file/filters/AbstractMarkerFilePresentFileListFilter.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+		return s -> s + suffix;
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public List<F> filterFiles(F[] files) {
+		if (files.length < 2) {
+			return Collections.emptyList();
+		}
+		final Set<String> candidates = Arrays.asList(files).stream()
```
Comment (artembilan) [inline spring-integration-file/src/main/java/org/springframework/integration/file/filters/AbstractMarkerFilePresentFileListFilter.java]: image

File: spring-integration-file/src/main/java/org/springframework/integration/file/filters/AbstractMarkerFilePresentFileListFilter.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+				.map(f -> getFilename(f))
```
Comment (artembilan) [inline spring-integration-file/src/main/java/org/springframework/integration/file/filters/AbstractMarkerFilePresentFileListFilter.java]: image

File: spring-integration-file/src/main/java/org/springframework/integration/file/filters/FileNameFileListFilter.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+ * A marker interface for {@link FileListFilter}s that indicates the filter
+ * examines file names.
+ *
+ * @param <F> the type of the object being filtered.
+ *
+ * @author Gary Russell
+ * @since 5.0
+ *
+ */
+public interface FileNameFileListFilter<F> extends FileListFilter<F> {
```
Comment (artembilan) [inline spring-integration-file/src/main/java/org/springframework/integration/file/filters/FileNameFileListFilter.java]: Gary, any strong requirements to have this extension?
Looks like it doesn't affect `AbstractMarkerFilePresentFileListFilter` unless there might be some stranger to inject an `AbstractMarkerFilePresentFileListFilter` into the `AbstractMarkerFilePresentFileListFilter`.
Thanks

File: spring-integration-file/src/main/java/org/springframework/integration/file/filters/AbstractMarkerFilePresentFileListFilter.java [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+		for (F file : files) {
+			if (this.filtersAndFunctions.entrySet().stream().anyMatch(entry -> {
+				F[] fileToCheck = (F[]) Array.newInstance(file.getClass(), 1);
+				fileToCheck[0] = file;
+				if (entry.getKey().filterFiles(fileToCheck).size() > 0) {
+					String markerName = entry.getValue().apply(getFilename(file));
+					return markerName != null && candidates.contains(markerName);
+				}
+				return false;
+			})) {
```
Comment (artembilan) [inline spring-integration-file/src/main/java/org/springframework/integration/file/filters/AbstractMarkerFilePresentFileListFilter.java]: Just nit-picking from my head. Would look better with a local variable for the `anyMatch()` result.

**Review by artembilan — APPROVED** [review]
That's all.
A couple minor concern and simple question.

I can merge with small polishing after your answer.

**Comment by artembilan** [comment]
@garyrussell ,

Any problems do not answer to my concerns and let me to merge then?

Thanks

**Comment by garyrussell** [comment]
I will do the requested polishing

**Pull request merged**